### PR TITLE
Improve block-selection strategy

### PR DESCRIPTION
**Pull request by Wizdave97**
Fixes https://github.com/paritytech/grandpa-bridge-gadget/issues/16

Round selection formula updated to match 

``` 
round_number =
      (1 - M) * session_start  
+    M * (best_beefy + NEXT_POWER_OF_TWO((best_grandpa - best_beefy + 1) / 2)) 
```

as outlined in https://github.com/paritytech/grandpa-bridge-gadget/blob/td-docs/docs/beefy.md#round-selection

Session boundary API that is used here is already implemented in https://github.com/paritytech/substrate/pull/10669

### CODE REVIEW

File: client/beefy/src/worker.rs [hunk 1]
```diff
@@ -461,13 +466,18 @@ where
 }
 
 /// Calculate next block number to vote on
-fn vote_target<N>(best_grandpa: N, best_beefy: N, min_delta: u32) -> N
+fn vote_target<N>(best_grandpa: N, best_beefy: N, session_start: N) -> N
 where
 	N: AtLeast32Bit + Copy + Debug,
 {
+	// if the session_start as a mandatory block does not have a beefy justification yet, we vote on
+	// it
```
Comment (acatangiu) [inline client/beefy/src/worker.rs:474]: nit:
```suggestion
	// if the session_start as a mandatory block does not have a beefy justification yet,
	// we vote on it
```

File: client/beefy/src/lib.rs [hunk 2]
```diff
@@ -126,8 +126,6 @@ where
 	pub signed_commitment_sender: BeefySignedCommitmentSender<B>,
 	/// BEEFY best block sender
 	pub beefy_best_block_sender: BeefyBestBlockSender<B>,
-	/// Minimal delta between blocks, BEEFY should vote for
-	pub min_block_delta: u32,
```
Comment (acatangiu) [inline client/beefy/src/lib.rs:130]: Let's not remove this. You can still use it on top of your changes as described in [beefy.md #L224](https://github.com/paritytech/grandpa-bridge-gadget/blob/ff8c141244cf62c61a4a1f646423712b0722820d/docs/beefy.md?plain=1#L224).

**Review by acatangiu — COMMENTED** [review]
Changes look good, only have some small comments. We also need https://github.com/paritytech/substrate/pull/10669 merged before this PR.

File: client/beefy/src/worker.rs [hunk 3]
```diff
@@ -146,18 +141,28 @@ where
...
 	C: Client<B, BE>,
-	C::Api: BeefyApi<B>,
+	C::Api: BeefyApi<B, NumberFor<B>>,
 {
 	/// Return `true`, if we should vote on block `number`
-	fn should_vote_on(&self, number: NumberFor<B>) -> bool {
+	fn should_vote_on(&self, header: &B::Header) -> bool {
+		let number = *header.number();
+		let at = BlockId::hash(header.hash());
+		let session_boundary = self.client.runtime_api().get_session_boundary(&at).ok();
```
[the remark this answers is not in the document]
Comment (Wizdave97) [reply client/beefy/src/worker.rs:155]: I'll keep that in mind.

**Comment by seunlanlege** [comment]
Bump

**Comment by acatangiu** [comment]
Alternative PR https://github.com/paritytech/substrate/pull/10882 merged

**Pull request closed without merging**